core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -201,8 +206,8 @@ class SYCLInternal {
   // An indirect kernel is one where the functor to be executed is explicitly
   // copied to USM memory before being executed, to get around the
   // trivially copyable limitation of SYCL.
-  using IndirectKernelMem = USMObjectMem<sycl::usm::alloc::shared>;
-  IndirectKernelMem m_indirectKernelMem;
+  using IndirectKernelMem = USMObjectMem<sycl::usm::alloc::device>;
```

Comment:
Yes it was. USM shared allocations induce page faults which really hurt performance. It's better, so long as the functor is never needed on the host, to write directly to device memory. This is particularly true in the context of this PR, which aims to make the functor data available on the device *before* the functor executes. Theoretically we can hide the data transfer time behind kernel execution.

Reply:
Otherwise, copying the kernel is blocking which this pull request tries to avoid as much as possible. We were discussing on slack whether choosing `device` is a good idea (since we must make sure that the original functor has not gone out-of-scope before the copy is finished) or not.